src/lager_logger_formatter.erl

```diff
@@ -0,0 +1,119 @@
+    Formatted = error_logger_lager_h:format_reason(Reason),
+    {errorContext, ErrorContext} = lists:keyfind(errorContext, 1, Report),
+    {offender, Offender} = lists:keyfind(offender, 1, Report),
+    case lists:keyfind(mod, 1, Offender) of
+        {mod, _Mod} ->
+            {pid, Pid} = lists:keyfind(pid, 1, Offender),
+            %% this comes from supervisor_bridge
+            {"Supervisor ~w had ~p ~p with reason ~s", [Name, Pid, ErrorContext, Formatted]};
+        false ->
+            {id, ChildID} = lists:keyfind(id, 1, Offender),
+            case lists:keyfind(pid, 1, Offender) of
+                {pid, Pid} ->
+                    {"Supervisor ~w had ~p ~p ~p with reason ~s", [Name, ChildID, Pid, ErrorContext, Formatted]};
+                false ->
+                    {"Supervisor ~w had ~p ~p with reason ~s", [Name, ChildID, ErrorContext, Formatted]}
+            end
+    end;
+report_cb(#{label := {application_controller, progress}, report := Report}) ->
+    case application:get_env(lager, suppress_application_start_stop, false) of
+        true -> {"", []};
+        false ->
+            {application, Name} = lists:keyfind(application, 1, Report),
+            {started_at, Node} = lists:keyfind(started_at, 1, Report),
+            {"Application ~w started on node ~w", [Name, Node]}
+    end;
+report_cb(#{label := {application_controller, exit}, report := Report}) ->
+    {exited, Reason} = lists:keyfind(exited, 1, Report),
+    case application:get_env(lager, suppress_application_start_stop) of
+        {ok, true} when Reason == stopped ->
+            {"", []};
+        _ ->
+            {application, Name} = lists:keyfind(application, 1, Report),
+            Formatted = error_logger_lager_h:format_reason(Reason),
+            {"Application ~w exited with reason: ~s", [Name, Formatted]}
+    end.
+%% TODO handle proc_lib crash
+
+format(#{msg := {report, _Report}, meta := Metadata} = Event, #{report_cb := Fun} = Config) when is_function(Fun, 1); is_function(Fun, 2) ->
+    format(Event#{meta => Metadata#{report_cb => Fun}}, maps:remove(report_cb, Config));
+format(#{level := _Level, msg := {report, Report}, meta := #{report_cb := Fun}} = Event, Config) when is_function(Fun, 1) ->
+    case Fun(Report) of
+        {"", []} -> "";
+        {Format, Args} when is_list(Format), is_list(Args) ->
+            format(Event#{msg => {Format, Args}}, Config)
+    end;
+format(#{level := Level, msg := {string, String}, meta := Metadata}, Config) ->
+    do_format(Level, String, Metadata, Config);
+format(#{level := Level, msg := {FmtStr, FmtArgs}, meta := Metadata}, Config) ->
+    Msg = lager_format:format(FmtStr, FmtArgs, maps:get(max_size, Config, 1024)),
+    do_format(Level, Msg, Metadata, Config).
+
+do_format(Level, Msg, Metadata, Config) ->
+    FormatModule = maps:get(formatter, Config, lager_default_formatter),
+    Timestamp = maps:get(time, Metadata),
+    MegaSecs = Timestamp div 1000000000000,
```

Comment:
Make these numbers macros to improve readability and maintenance sanity, please.